File: backend/app/core/password_policy.py
import re

PASSWORD_POLICY_MESSAGE = (
    "Password must be at least 8 characters and include an uppercase letter, "
    "lowercase letter, number, and special character."
)
PASSWORD_MAX_LENGTH = 72

_UPPERCASE_PATTERN = re.compile(r"[A-Z]")
_LOWERCASE_PATTERN = re.compile(r"[a-z]")
_NUMBER_PATTERN = re.compile(r"\d")
_SPECIAL_PATTERN = re.compile(r"[^A-Za-z0-9]")


def validate_password_strength(password: str) -> str:
    value = str(password or "")

    if len(value) > PASSWORD_MAX_LENGTH:
        raise ValueError(f"Password must be {PASSWORD_MAX_LENGTH} characters or fewer.")

    if (
        len(value) < 8
        or not _UPPERCASE_PATTERN.search(value)
        or not _LOWERCASE_PATTERN.search(value)
        or not _NUMBER_PATTERN.search(value)
        or not _SPECIAL_PATTERN.search(value)
    ):
        raise ValueError(PASSWORD_POLICY_MESSAGE)

    return value

File: frontend/src/utils/passwordPolicy.js
export const PASSWORD_POLICY_MESSAGE = 'Password must be at least 8 characters and include an uppercase letter, lowercase letter, number, and special character.'

const PASSWORD_RULES = [
  { label: 'At least 8 characters', test: (password) => password.length >= 8 },
  { label: 'Uppercase letter', test: (password) => /[A-Z]/.test(password) },
  { label: 'Lowercase letter', test: (password) => /[a-z]/.test(password) },
  { label: 'Number', test: (password) => /\d/.test(password) },
  { label: 'Special character', test: (password) => /[^A-Za-z0-9]/.test(password) }
]

export function getPasswordChecks(password) {
  const value = String(password || '')
  return PASSWORD_RULES.map((rule) => ({
    label: rule.label,
    ok: rule.test(value)
  }))
}

export function isPasswordPolicySatisfied(password) {
  return getPasswordChecks(password).every((check) => check.ok)
}
